File: src/analyze_scf_asymmetry.py
# m_matschiner Mon Feb 5 18:03:41 CET 2024

# Import libraries.
import sys
import numpy as np
from scipy.stats import binom

# Define method for two-sided binomial test.
def two_sided_binomial_test(x, n, p):
    """
    Perform a two-sided binomial test.
    
    Parameters:
    - x: the number of successes
    - n: the number of trials
    - p: the probability of success
    
    Returns:
    - p_value: the p-value of the test
    """
    # Calculate the p-value using the cumulative distribution function (CDF)
    p_value = min(binom.cdf(x, n, p), 1 - binom.cdf(x - 1, n, p)) * 2
    
    return p_value

# Get the file name from the command-line argument
table_file_name = sys.argv[1]

# Open the input file.
with open(table_file_name,'r') as file:
    lines = file.readlines()

    # Read the input file.
    p1p3 = int(float(lines[1].split()[8]))
    p2p3 = int(float(lines[1].split()[9]))
    x = p1p3
    n = p1p3 + p2p3
    p = 0.5
    
    # Calculate the D-statistic.
    if p1p3 + p2p3 == 0:
        d = 0
    else:
        d = (p1p3 - p2p3) / (p1p3 + p2p3)
        
    # Calculate the p-value.
    p_value = two_sided_binomial_test(x, n, p)
        
    # Report d-statistic and p-value.
    print(str(d) + "\t" + str(p_value))

File: src/analyze_scf_asymmetry.rb
# m_matschiner Fri Feb 2 16:52:54 CET 2024

# See https://en.wikipedia.org/wiki/Binomial_coefficient
def n_choose_k(n, k)
	if k > n
		puts "ERROR: k should never be larger than n!"
		exit 1
	elsif k == 0
		return 1
	elsif k > n/2
		return n_choose_k(n, n-k)
	end
	r = 1
	1.upto(k) do |j|
		r = r * (n + 1 - j) / j
	end
	r
end

# Get the command-line arguments.
table_file_name = ARGV[0]

# Read the input file.
table_file = File.open(table_file_name, "r")
lines = table_file.readlines

# Get the numbers of sites supporting (p1,p2), (p1,p3), and (p2,p3)
p1p2 = lines[1].split[7].to_f
p1p3 = lines[1].split[8].to_f
p2p3 = lines[1].split[9].to_f

# Calculate the d statistic.
if p1p3 + p2p3 == 0
	d = 0
else
	d = (p1p3 - p2p3)/(p1p3 + p2p3)
end

# Test if the imbalance between discordant trees is greater than expected by chance.
pr = 0
n = p1p3.to_i + p2p3.to_i
k = (p2p3.to_i+1)
k.times do |i|
	pr += Math.exp(Math.log(n_choose_k(n,i)) + Math.log(0.5**i) + Math.log((1-0.5)**(n-i)))
end
pr = pr * 2 # Make the test two-tailed instead of one-tailed.
pr = 1.0 if pr > 1.0 # This is possible when both alternative topologies are equally frequent.

# Report the D-statistic and the calculated p-value.
puts "#{d}\t#{pr}"
